File: src/main/java/toby/jpa/persistence/impl/UserPersistenceImpl.java
package toby.jpa.persistence.impl;

import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;
import toby.jpa.dto.MusicDto;
import toby.jpa.dto.UserDto;
import toby.jpa.persistence.IUserPersistence;
import toby.jpa.service.IMusicFileService;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import javax.persistence.Query;
import java.util.List;

@Repository
@Transactional
public class UserPersistenceImpl implements IUserPersistence {

    private final IMusicFileService musicFileService;

    UserPersistenceImpl(IMusicFileService musicFileService) {
        this.musicFileService = musicFileService;
    }

    @PersistenceContext
    protected EntityManager em;

    public EntityManager getEntityManager() {
        return em;
    }

    public void setEntityManager(EntityManager entityManager) {
        this.em = entityManager;
    }


    @Override
    @SuppressWarnings("unchecked")
    public List<UserDto> listGuildUsers(Long guildId) {
        Query q = em.createNamedQuery("UserDto.getGuildAll", UserDto.class);
        q.setParameter("guildId", guildId);
        return q.getResultList();
    }

    @Override
    public UserDto createNewUser(UserDto userDto) {
        createMusicFileEntry(userDto);
        UserDto databaseUser = em.find(UserDto.class, userDto);
        UserDto result = (databaseUser == null) ? persistConfigDto(userDto) : databaseUser;

        return result;
    }

    private void createMusicFileEntry(UserDto userDto) {
        MusicDto musicDto = new MusicDto(userDto.getDiscordId(), userDto.getGuildId(), null, null);
        musicFileService.createNewMusicFile(musicDto);
        em.flush();
    }

    @Override
    public UserDto getUserById(Long discordId, Long guildId) {
        Query userQuery = em.createNamedQuery("UserDto.getById", UserDto.class);
        userQuery.setParameter("discordId", discordId);
        userQuery.setParameter("guildId", guildId);
        return (UserDto) userQuery.getSingleResult();
    }

    @Override
    public UserDto updateUser(UserDto userDto) {
        UserDto dbUser = getUserById(userDto.getDiscordId(), userDto.getGuildId());
        MusicDto musicFileById = musicFileService.getMusicFileById(userDto.getMusicDto().getId());
        MusicDto requestMusicDto = dbUser.getMusicDto();
        if (!requestMusicDto.equals(musicFileById)) {
            musicFileService.updateMusicFile(requestMusicDto);
        }

        if (!userDto.equals(dbUser)) {
            em.merge(userDto);
            em.flush();
        }

        return userDto;
    }

    @Override
    public void deleteUser(UserDto userDto) {
        em.remove(userDto.getMusicDto());
        em.remove(userDto);
        em.flush();
    }

    @Override
    public void deleteUserById(Long discordId, Long guildId) {

        Query userQuery = em.createNamedQuery("UserDto.deleteById");
        userQuery.setParameter("discordId", discordId);
        userQuery.setParameter("guildId", guildId);
        userQuery.executeUpdate();

    }

    private UserDto persistConfigDto(UserDto userDto) {
        em.persist(userDto);
        em.flush();
        return userDto;
    }
}

File: src/main/kotlin/toby/jpa/persistence/impl/UserPersistenceImpl.kt
package toby.jpa.persistence.impl

import jakarta.persistence.EntityManager
import jakarta.persistence.PersistenceContext
import jakarta.persistence.Query
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional
import toby.jpa.dto.UserDto
import toby.jpa.persistence.IUserPersistence
import toby.jpa.service.IMusicFileService

@Repository
@Transactional
open class UserPersistenceImpl internal constructor(private val musicFileService: IMusicFileService) : IUserPersistence {
    @PersistenceContext
    lateinit var entityManager: EntityManager


    override fun listGuildUsers(guildId: Long?): List<UserDto?> {
        val q: Query = entityManager.createNamedQuery("UserDto.getGuildAll", UserDto::class.java)
        q.setParameter("guildId", guildId)
        return q.resultList as List<UserDto?>
    }

    override fun createNewUser(userDto: UserDto): UserDto {
        createMusicFileEntry(userDto)
        val databaseUser = entityManager.find(UserDto::class.java, userDto)

        return if ((databaseUser == null)) persistUserDto(userDto) else databaseUser
    }

    private fun createMusicFileEntry(userDto: UserDto) {
        userDto.musicDtos.forEach { musicFileService.createNewMusicFile(it) }
        entityManager.flush()
    }

    override fun getUserById(discordId: Long?, guildId: Long?): UserDto? {
        val userQuery: Query = entityManager.createNamedQuery("UserDto.getById", UserDto::class.java)
        userQuery.setParameter("discordId", discordId)
        userQuery.setParameter("guildId", guildId)
        return runCatching { userQuery.singleResult as UserDto? }.getOrNull()
    }

    override fun updateUser(userDto: UserDto): UserDto {
        val dbUser = getUserById(userDto.discordId, userDto.guildId)
        userDto.musicDtos.forEach {
            val musicFileById = musicFileService.getMusicFileById(it.id!!)
            val requestMusicDtos = dbUser?.musicDtos
            if (requestMusicDtos?.contains(musicFileById) == true) {
                musicFileService.updateMusicFile(it)
            }

        }

        if (userDto != dbUser) {
            entityManager.merge(userDto)
            entityManager.flush()
        }

        return userDto
    }

    override fun deleteUser(userDto: UserDto) {
        entityManager.remove(userDto.musicDtos)
        entityManager.remove(userDto)
        entityManager.flush()
    }

    override fun deleteUserById(discordId: Long?, guildId: Long?) {
        deleteAssociatedMusicFilesForUserAndGuild(discordId, guildId)

        val userQuery = entityManager.createNamedQuery("UserDto.deleteById")
        userQuery.setParameter("discordId", discordId)
        userQuery.setParameter("guildId", guildId)
        userQuery.executeUpdate()
    }

    private fun deleteAssociatedMusicFilesForUserAndGuild(discordId: Long?, guildId: Long?) {
        val musicFileQuery = entityManager.createNamedQuery("MusicDto.deleteByUser")
        musicFileQuery.setParameter("discordId", discordId)
        musicFileQuery.setParameter("guildId", guildId)
        musicFileQuery.executeUpdate()
    }

    private fun persistUserDto(userDto: UserDto): UserDto {
        entityManager.persist(userDto)
        entityManager.flush()
        return userDto
    }
}
